core/include/keyman/keyman_core_api.h

```diff
@@ -526,6 +558,128 @@ enum km_core_action_type {
   KM_CORE_IT_MAX_TYPE_ID
 };
 
+/*
+```
+### Actions
+This structure provides the results of processing a key event to the Platform layer and
+should be processed by the Platform layer to issue commands to the os text
+services framework to transform the text store in the Client Application, among
+other actions.
+
+This API replaces the Action items APIs, which is now deprecated and will be
+removed in the future.
+```c
+*/
+
+typedef struct {
+  // number of codepoints (not codeunits!) to delete from app context, 0+
+  int delete_back;
+
+  // null-term string of characters to insert into document
+  km_core_usv* output;
+
+  // list of options to persist, terminated with KM_CORE_OPTIONS_END
+  km_core_option_item* persist_options;
+
+  // issue a beep
+  bool do_alert;
+
+  // emit the input keystroke to the application, unmodified?
+  bool emit_keystroke;
+
+  // -1=unchanged, 0=off, 1=on
+  int new_caps_lock_state;
+} km_core_actions;
+
+/*
+```
+### `km_core_state_get_actions`
+##### Description:
+Returns a pointer to an actions object which details all the actions
+that the Platform layer must take after a keystroke. The `delete_back`
+action must be performed before the `output` action, but the other
+actions may be performed in any order.
+##### Return:
+A pointer to a `km_core_actions` object, which must be freed with
+`km_core_actions_dispose`.
+##### Parameters:
+- __state__: An opaque pointer to a state object.
+
+```c
+*/
+KMN_API
+km_core_actions*
+km_core_state_get_actions(
+  km_core_state const *state
+);
+
+/*
+```
+### `km_core_actions_dispose`
+##### Description:
+Free the allocated memory belonging to an actions object previously
+returned by `km_core_state_get_actions`.
+##### Parameters:
+- __actions__: A pointer to the actions object to be disposed of.
+
+```c
+*/
+KMN_API
+km_core_status
+km_core_actions_dispose(
+  km_core_actions* actions
+);
+
+/*
+```
+### `km_core_state_context_validate`
+##### Description:
+Determines if the input context has changed, and if so, resets
+the internal cached context (including markers), to the new
+context string.
+
```

Comment:
I'm a little confused by this description: the idea of the context changing but the API name is validate. You say that it 'resets', but I think of a reset as clearing the context not setting it to a new value. The part about including markers seems a little ambiguous, as if the markers could also be provided in the input, but it should never have any.

Maybe it would be clearer if it said that it was *replacing* the internal cached context *with* the input context string. (If I'm understanding that correctly.)

An API named validate doesn't sound like it is going to change state. So if it has the possibility of replace the current context in the core, then I think it needs to reflect that.

Reply:
The question is, when should the caller call this? Should be named something such as `update_state_with_new_context` ?

Reply:
@srl the caller would now call this on every keystroke instead of `km_core_context_set`.  Before when the platform was managing its own version of the local cache (with markers) it could strip out the markers and see if the context was still the same. For example, the user had not moved the cursor to a new position.  
I agree that `validate` is not a good name as it doesn't suggest it is doing any mutation. 
Either the api has a call that just does a validation/match only and then if returns `false` we call `km_core_context_set`.
Or we have a single function that provides that functionality? 
`km_km_core_context_confirm_match_or_set`   -- a bit long maybe @sgschantz  will have a better name.

Reply:
Agree that the function names are somewhat difficult. Am going to refactor this into two functions I think, although that makes usage on Engine side slightly more difficult.